sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py

```diff
@@ -0,0 +1,378 @@
+
+                # update moments accountant
+                alphas = alphas + moments_acc(self.num_teachers, votes, noise_multiplier, l_list)
+
+                loss_s = criterion(output, predictions.float().to(self.device))
+
+                loss_s.backward()
+                optimizerS.step()
+                
+                print ('iterator {i}, student discriminator loss is {j}'.format(i=t_3, j=loss_s))
+
+            # train generator
+            fakez = torch.normal(mean=mean, std=std)
+            condvec = self.cond_generator.sample(self.batch_size)
+
+            if condvec is None:
+                c1, m1, col, opt = None, None, None, None
+            else:
+                c1, m1, col, opt = condvec
+                c1 = torch.from_numpy(c1).to(self.device)
+                m1 = torch.from_numpy(m1).to(self.device)
+                fakez = torch.cat([fakez, c1], dim=1)
+
+            fake = self.generator(fakez)
+            fakeact = self._apply_activate(fake)
+
+            if c1 is not None:
+                y_fake = student_disc(torch.cat([fakeact, c1], dim=1))
+            else:
+                y_fake = student_disc(fakeact)
+
+            if condvec is None:
+                cross_entropy = 0
+            else:
+                cross_entropy = self._cond_loss(fake, c1, m1)
+
+
+            if self.loss=='cross_entropy':
+                label_g = torch.full((int(self.batch_size/self.pack),), REAL_LABEL, dtype=float, device=self.device)
+                #label_g = torch.full(int(self.batch_size/self.pack,),1,device=self.device)
+                loss_g = criterion(y_fake, label_g.float())
+                loss_g = loss_g + cross_entropy
+            else:
+                loss_g = -torch.mean(y_fake) + cross_entropy
+
+            
+            optimizerG.zero_grad()
+            loss_g.backward()
+            optimizerG.step()
+                
+                
+            print ('generator is {}'.format(loss_g.detach().cpu()))
+
+            eps = min((alphas - math.log(self.delta)) / l_list)
+
+
+    def generate(self, n):
+        self.generator.eval()
+
+        steps = n // self.batch_size + 1
+
+        data = []
+
+        for i in range(steps):
+            mean = torch.zeros(self.batch_size, self.embedding_dim)
+            std = mean + 1 
+            fakez = torch.normal(mean=mean, std=std).to(self.device)
+
+            condvec = self.cond_generator.sample_zero(self.batch_size)
+            if condvec is None:
+                pass
+            else:
+                c1 = condvec
+                c1 = torch.from_numpy(c1).to(self.device)
+                fakez = torch.cat([fakez, c1], dim=1)
+
+
+            fake = self.generator(fakez)
+            fakeact = self._apply_activate(fake)
+            data.append(fakeact.detach().cpu().numpy())
+        
+        data = np.concatenate(data, axis=0)
+        data = data[:n]
+
+        return self.transformer.inverse_transform(data, None)
+
```

Comment:
please check for new lines. There should be at most 1 or 2 new lines for spacing. More than that should be reduced